rsync_backup.sh

```diff
@@ -33,24 +33,72 @@ check_if_directory() {
   fi
 }
 
-count_files() {
-  path_passed_in=$1
-  find "${path_passed_in%/}" -type f 2> /dev/null | wc -l
-
-  if [ ! -d "$path_passed_in" ]; then
-    echo "There are NO files at this path" "$path_passed_in"
-    exit 1
-  fi
-}
-
 count_files_dirs_etc() {
+  # Counts files, directories, symlinks, etc. in path passed in.
   path_passed_in=$1
   find "${path_passed_in%/}" 2> /dev/null | wc -l
+}
 
-  if [ ! -d "$path_passed_in" ]; then
-    echo "There are NO files, directories, symlinks, etc. at this path" "$path_passed_in"
-    exit 1
-  fi
+print_number_of_files() {
+  local source_or_dest=$1
+  local number_of_files=$2
+
+  echo "# of files, dirs, symlinks, etc. in: " "$source_or_dest" "is: " "$number_of_files"
+}
+
+do_backup_for_2_targets() {
+  # For Case 1, Documents & Case2, PHOTOS.
+  # Note: Passes for Documents case, Case 1:
+  # source_path_Documents="/Users/kimlew/Documents"
+  # dest_red="/Volumes/ToshibaRD/"
+  # dest_blue="/Volumes/ToshibaBL/"
+  # Called: do_backup "$source_path_Documents" "Documents" "$dest_red" "Red Toshiba Hard Drive" "$dest_blue" "Blue Toshiba Hard Drive"
+
+  local source_path=$1
+  local source_name=$2
+  
+  local dest_path_red=$3
+  local dest_name_red=$4
+
+  local dest_path_blue=$5
+  local dest_name_blue=$6
+  
+  echo "BACKUP in progress..."
+  echo
+
+  # -a, --archive - archive mode; same as -rlptgoD (no -H). -a implies -r.
+  # -v is verbose vs. -q, --quiet - to suppress non-error messages.
+  # > dry-run_Documents_to_RedHD.txt \
+  rsync -avi --progress --stats --exclude={'.DocumentRevisions-V100','.TemporaryItems','.Spotlight-V100','.Trashes','.fseventsd'} \
+  "$source_path" "$dest_path_red" \
+  > backup_Documents_to_RedHD.txt \
+  && echo "BACKUP DONE of $source_name -> $dest_name_red."
+
+  rsync -avi --progress --stats --exclude={'.DocumentRevisions-V100','.TemporaryItems','.Spotlight-V100','.Trashes','.fseventsd'} \
+  "$source_path" "$dest_path_blue" \
+  > backup_Documents_to_BlueHD.txt \
+  && echo "BACKUP DONE of $source_name -> $dest_name_blue."
+  echo
+}
+do_backup_for_1_target() {
+  local source_path=$1
+  local source_name=$2
+  
+  local dest_path=$3
+  local dest_name=$4
+  
+  echo "BACKUP in progress..."
+  echo
+
+  # -a, --archive - archive mode; same as -rlptgoD (no -H). -a implies -r.
+  # -v is verbose vs. -q, --quiet - to suppress non-error messages.
+  # > dry-run_Documents_to_RedHD.txt \
+  rsync -avi --progress --stats --exclude={'.DocumentRevisions-V100','.TemporaryItems','.Spotlight-V100','.Trashes','.fseventsd'} \
+  "$source_path" "$dest_path" \
+  > backup_USB_to_Hard_Drive.txt \
+  && echo "BACKUP DONE of $source_name -> $dest_name."
+  echo
+  # TODO: Change .txt to increase in number when script runs? How else 2 files?
 }
```

Comment:
This is looking pretty good, but you might have noticed that the body of `do_backup_for_2_targets` is very similar to the body of `do_backup_for_1_target`. Right now, calling `do_backup_for_2_targets` is very similar to calling `do_backup_for_1_target` twice, except

- you would print `BACKUP in progress...` twice
- you currently hard-code the names of the files you redirect the output of `rsync` to

If you made two little changes, you could replace one call of `do_backup_for_2_targets` with two calls of `do_backup_for_1_target`:


1. change the 'in progress' message  in `do_backup_for_1_target` to be something like
    ```
    echo "BACKUP IN PROGRESS for ${source_name} -> ${dest_name}."
    ```
2. add a fifth parameter to `do_backup_for_1_target`, the name of the file you'll redirect the `rsync` output to

Then instead of calling

```
do_backup_for_2_targets "$source_path_Documents" "Documents" "$dest_red" "Red Toshiba Hard Drive" "$dest_blue" "Blue Toshiba Hard Drive"
```
you would have two calls that look like
```
do_backup_for_1_target "$source_path_Documents" "Documents" \
  "$dest_red" "Red Toshiba Hard Drive" backup_Documents_to_RedHD.txt
do_backup_for_1_target "$source_path_Documents" "Documents" \
  "$dest_blue" "Blue Toshiba Hard Drive" backup_Documents_to_BlueHD.txt
```
I have used continuation backslashes to break up the long command line.

This way you could remove `do_backup_for_2_targets` and avoid having three different places where you have the `rsync` command with its long `--exclude={...}` argument. You might find that eventually you want to add more stuff to that list of excluded files and directories and it would be easy to forget to update one of the three places.

Reply:
@kcharter I did try this as you initially suggested, but the message with totals printed twice which I do not want.
I will keep this in mind in case a solution reveals itself.

Reply:
Sounds good @kimlee